### PR TITLE
Generalize VM benchmark suite / add profiling / separate code & execution logic

### PR DESCRIPTION
**Pull request by holgerd77**
API changes within this PR:

- New `RunBlockResult` type and added result parameters in 3cbae464ab96dba93008f29b7a835163222a92b3

~Update: This PR contains two **VM bug fixes** from @jochem-brouwer which should not be forgotten on release notes~

- ~Fix sstore gas cost calculation if two txs go to same contract in the same block, a015926ceced6dc87f94f2a041a8ac351f0085ad~
- ~Fix Istanbul sstoreCleanRefundEIP2200 gas cost in case of wrongly padded buffers, ae2cd6163c6452be9b198df14c41b8f77255fe8d~

Update: these fixes have been extracted to #870 

---

This PR generalizes the VM benchmarking setup to set some ground for expanding with other code snippets for benchmarking over time and allow for flexible combination of these different benchmarks on CI runs.

In particular it does the following:
- Move benchmarks from `scripts/benchmarks` to `benchmarks` for better visibility
- Separate the benchmark setup code from the to be benchmarked code to allow for...
  - Simpler execution and debugging of the respective code parts
  - Run the code in a profiling context
- Added profiling with 0x as an alternative run mode to `package.json` (`npm run benchmarks`, `npm run profiling`), allowing to run code snippets either in a benchmarking or profiling context
- More flexible and expandable API and file structure to be more flexible on adding new benchmark code snippets for other parts of the VM (StateManager, Cache,...)

I still have got the following problems:

When running in a non-benchmark context (so e.g. for profiling with `npm run build:benchmarks && npm run profiling -- mainnetBlocks:5`) I got a `StackUnderflow` error, I have no idea why this is:

```shell
(node:49853) UnhandledPromiseRejectionWarning: Error: Stack underflow
    at SecureTrie._updateNode (/ethereumjs-vm/node_modules/merkle-patricia-tree/dist/baseTrie.js:295:19)
    at SecureTrie.put (/ethereumjs-vm/node_modules/merkle-patricia-tree/dist/baseTrie.js:142:24)
    at async SecureTrie.put (/ethereumjs-vm/node_modules/merkle-patricia-tree/dist/secure.js:79:13)
(node:49853) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 36)
(node:49853) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

I get highly different benchmark output on the `master` version:

```shell
Total number of blocks in data set: 15
Number of blocks to sample: 15
Block 9422905 x 1,870 ops/sec ±6.12% (81 runs sampled)
Block 9422906 x 1,939 ops/sec ±3.44% (86 runs sampled)
Block 9422907 x 1,559 ops/sec ±14.86% (69 runs sampled)
Block 9422908 x 1,772 ops/sec ±4.08% (80 runs sampled)
```

And the runs on this branch:

```shell
Benchmarking started...
Running 'mainnetBlocks':
Total number of blocks in data set: 15
Number of blocks to sample: 5
Block 9422905 x 39,468 ops/sec ±7.14% (52 runs sampled)
Block 9422906 x 24,251 ops/sec ±72.81% (52 runs sampled)
Block 9422907 x 43,236 ops/sec ±4.03% (65 runs sampled)
Block 9422908 x 39,880 ops/sec ±12.02% (59 runs sampled)
```

Would be happy if someone would have a second look here.

### CODE REVIEW

**Comment by jochem-brouwer** [comment]
Very suspicious that there's a 20x+ speed increase. I will take a look at this `StackUnderflow` error.

**Comment by jochem-brouwer** [comment]
Stack underflow happens twice during a `SSTORE`:

Address: `57831a0c76ba6b4fdcbadd6cb48cb26e8fc15e93`, storage slot: `4f3ce86493742328fffe784b75c68c34f495f416d51e49356e159bac9092710`

Address: `a5d7d68d7975e89feb240f42fed1d77bb71b1caf`, storage slot: `9267d898f406d2d0c108c7933c558e54642b0b339379caa2fe77a4f8be716a53`

**Comment by ryanio** [comment]
> @holgerd77 Should the pre-state be complete for the block itself (such that the block itself can be ran in isolation?). It looks like these states are missing from pre-state at the relevant blocks (but they are initialized in earlier blocks).

ohh that could answer why sina set up all of the pre-state before running the first block, @holgerd77 and i were previously wondering why it was set up like that since it was quite slow in practice.

if possible it would be nice to have a complete pre-state for each block for them to be tested in isolation without dependency on running other blocks, but i'm open to other ideas.

**Comment by jochem-brouwer** [comment]
I actually messed up the block numbers, the state looks to be available in the pre-state. However I think there might be a problem in `vm.copy()`, the state manager thinks it is on Petersburg (should be on MuirGlacier).

It is also worth figuring out why this promise is not caught and what causes it in the first place, so will focus on that first.

**Comment by jochem-brouwer** [comment]
For reference: just put a (temporary) MPT state test here. This initializes the StateManager to the preState of block 9422905. If you try to edit the value, it throws. (Investigating...)

**Comment by jochem-brouwer** [comment]
Yeah I found the root of the problem. Every storage slot defaults to 0. For some reason the state is not loaded the right way. I don't understand why as the code seems to be correct.

How is the speed of the benchmark generated? If it depends on the expectation that block gas is being used then this would explain why the ops/s is suddenly much faster: Most of the gas is actually not being used as all storage slots default to 0 and therefore most execution halts quickly (VM follows in most cases another path that the actual execution path on the chain).

**Comment by jochem-brouwer** [comment]
@holgerd77 I have rewritten `util.ts`. I can imagine that you wrote `util.ts` like this (use low-level functions to setup preState), but this will break if we modify anything in `StateManager`. I tried getting it to work as-is but for some reason it looks like state is getting lost at some point. I understand if this is done to not influence the benchmark so much, but can't we disable the benchmark until pre-state is loaded and then enable it? 

Will add receipts/gasUsed/logs check.

**Comment by jochem-brouwer** [comment]
Found some more issues initializing the state: Common of block not initialed correctly when calling `blockFromRPC` (it uses the default Common which yields wrong upfront costs for the transactions, as the gas prices for txDataNonZero has been changed in Istanbul).

Also, annoying: the pre-state has some storage slots which are initialized to a zero value. These are written to contract storage (though these should not be written!). In gas calculation, VM thinks that these values thus exist in storage (and if you SSTORE them it deducts 5000 gas instead of 20000). Linked: #729. Will push fix.

**Comment by jochem-brouwer** [comment]
[Fixed](https://github.com/ethereumjs/ethereumjs-vm/pull/853/commits/a015926ceced6dc87f94f2a041a8ac351f0085ad) a possible VM bug: we do not clear up the original contract storage after we have ran a Tx; this *did* run into gas/receipt issues in the mainnet block.

Another issue is that a transaction needs access to a past `BLOCKHASH` which we currently do not seem to provide.

**Comment by jochem-brouwer** [comment]
Another problem has to do with the Istanbul storage calculation: we try to write (transaction `0x6e65041a9f03830741989d05be737adb23332205cbc7e77981fca996b5a115b0`);

```
<Buffer 00 00 00 00 00 00 00 00 00 00 00 00 7d 34 55 42 1b bc 5e d5 34 a8 3c 88 fd 80 38 7d c8 27 13 92>
```

To a slot where original value is: 

```
<Buffer 7d 34 55 42 1b bc 5e d5 34 a8 3c 88 fd 80 38 7d c8 27 13 92>
```

These buffers are supposed to be the same value, but due to their length difference [this equality check](https://github.com/ethereumjs/ethereumjs-vm/blob/428ae935de4d01198354d826867303068786bb4b/packages/vm/lib/evm/opFns.ts#L1070) returns `false` and therefore we are not getting a refund while we should get a refund.

**Comment by jochem-brouwer** [comment]
The blocks are now running and they pass the receiptsTrie/logsBloom/gasUsed checks so we can be fairly certain that VM runs correctly on those blocks. The fix I pushed is rather ugly, we should be more consistent. Especially this variable-level-byte-length stuff is rather confusing and is prone to errors.

On a side note, very weird that ethereum/tests did not catch these bugs!

**Comment by holgerd77** [comment]
Hi @jochem-brouwer thanks for the in-depth analysis of the problems and the fixes here, have updated the initial post so that these won't get lost on the VM release notes.

Would you generally regard this PR now as ready-for-review? Will wait for #863 to be reviewed/merged and then do a rebase here so that we can finish (sorry to add to the merge-dependency-chain for #849 here 🙂 )

**Comment by jochem-brouwer** [comment]
@holgerd77 ah yeah it is ready for review, thought you wanted to add  or change some more stuff 😃

File: packages/vm/benchmarks/mainnetBlocks.ts
Comment (ryanio) [inline packages/vm/benchmarks/mainnetBlocks.ts:23]: small suggestion: compare buffers directly
```suggestion
    !result.receiptRoot.equals(block.header.receiptTrie)
```

File: packages/vm/benchmarks/run.ts
Comment (ryanio) [inline packages/vm/benchmarks/run.ts:20]: this is nice, but i don't see it being output anywhere

Comment (holgerd77) [reply packages/vm/benchmarks/run.ts:20]: Have corrected (next push).

**Comment by holgerd77** [comment]
Ok, have rebased this and did some fixes.

Results still seem to be pretty unreliable and this will need some further investigation, here is some example output:

```shell
Benchmarking started...
Running 'mainnetBlocks':
Total number of blocks in data set: 15
Number of blocks to sample: 5
Block 9422905 x 44,786 ops/sec ±36.05% (57 runs sampled)
Block 9422906 x 49,694 ops/sec ±4.46% (67 runs sampled)
Block 9422907 x 44,755 ops/sec ±12.26% (61 runs sampled)
Block 9422908 x 15,700 ops/sec ±131.05% (60 runs sampled)
Block 9422909 x 12,869 ops/sec ±131.21% (61 runs sampled)
Benchmark run finished.
```

@jochem-brouwer you might want to take your fixes out of here and do separate PRs on this - this would be the cleaner solution anyhow. Then I can rebase on top here once we merge these separate PRs and you can use the fixes for your subsequent work (eventually directly try to improve on the byte-length code - you might at least be able to use `Util` functionality here and we should encapsulate and/or generalize better - think we can't merge this yet in its current form). This PR will likely take some more time to get to some merge-ready state.

**Comment by holgerd77** [comment]
(so the fixes directly applied to the VM code, not the benchmark fixes)

**Comment by jochem-brouwer** [comment]
I will push a branch with only the VM bugfixes, I will leave it to you to then remove those commits from this PR (and just add those when you rebase on master if these bugfixes are merged).

**Comment by jochem-brouwer** [comment]
Turned out that it was a bit more complex than it was, have removed the bugfix commits. Should be fine now. (If there are any problems I have a local backup before my force push 😅 )

File: packages/vm/lib/runBlock.ts
Comment (holgerd77) [inline packages/vm/lib/runBlock.ts:202]: @jochem-brouwer This is a cool change but changes like these absolutely need to be documented. This is just too dangerous otherwise that some inconsistencies slip through or bugs are introduced, especially since this is part of the end user API.

Are we sure here this is now what we want to expose as a `RunBlockResult`, so that we can keep this stable mid-term? 🤔  I haven't actually put a lot of thought yet on this question.

**Comment by holgerd77** [comment]
Rebased this now containing the fixes from #870

File: packages/vm/lib/runBlock.ts
Comment (ryanio) [reply packages/vm/lib/runBlock.ts:195]: i added a changelog entry for this: https://github.com/ethereumjs/ethereumjs-vm/pull/853/commits/6d21c41072148acd3c2b9db95731bc2b5cb81648

**Comment by ryanio** [comment]
i just rebased on master, and did a few things: updated docs for new script format, added changelog entry for runBlock result, refactored `verifyResult` out to `./utils`, and some other misc. code cleanup.

**Comment by ryanio** [comment]
@holgerd77 I finally found the reason for the ops/sec increase! I had to step through every single change in the PR but it turns out it was the `bind` here, I did not expect it:

```typescript
suite.add(`Block ${blockNumber}`, onAdd.bind(vm, block, receipts))
```

changing it to the below fixed it:

```typescript
suite.add(`Block ${blockNumber}`, async () => {
  await onAdd(vm, block, receipts)
})
```

Result:

```
Benchmarking started...
Running 'mainnetBlocks':
Total number of blocks in data set: 15
Number of blocks to sample: 10
Block 9422905 x 2,052 ops/sec ±2.05% (89 runs sampled)
Block 9422906 x 2,053 ops/sec ±1.43% (88 runs sampled)
Block 9422907 x 1,906 ops/sec ±3.44% (86 runs sampled)
Block 9422908 x 1,863 ops/sec ±3.77% (85 runs sampled)
Block 9422909 x 1,786 ops/sec ±2.80% (81 runs sampled)
Block 9422910 x 1,937 ops/sec ±1.62% (88 runs sampled)
```

I since renamed `onAdd` to `runBlock` to make it more descriptive/self-explanatory.

Ok, finally ready for review, this one has been on my list for a while! I think it looks great. I also included some fixes for the trie benchmarks.

**Review by ryanio — DISMISSED** [review]

**Comment by holgerd77** [comment]
Great! 😄  Think we'll likely wait here to merge until we have merged #1048 so that we can then rebase the small PR on top of the big one.

**Comment by jochem-brouwer** [comment]
Can someone explain to me why this `bind` thing results in such a massive performance decrease?

**Comment by ryanio** [comment]
> Can someone explain to me why this `bind` thing results in such a massive performance decrease?

I think it was a performance increase actually, my hunch is that it wasn't awaiting the async func so it was able to run it 10-20x faster 😅

**Comment by holgerd77** [comment]
Have merged by admin-merge here since this was already reviewed by @ryanio.

**Pull request merged**
